File: packages/sdk/src/crypto.ts
import * as ed from "@noble/ed25519";
import { sha512 } from "@noble/hashes/sha512";
import { sha256 } from "@noble/hashes/sha256";
import type { KeyPair } from "./types.js";

// Required for @noble/ed25519 in Node/browser
ed.etc.sha512Sync = (...m) => sha512(ed.etc.concatBytes(...m));

const b64u = {
  encode: (bytes: Uint8Array): string =>
    Buffer.from(bytes).toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, ""),
  decode: (s: string): Uint8Array => {
    const pad = s.length % 4 === 0 ? "" : "=".repeat(4 - (s.length % 4));
    return new Uint8Array(Buffer.from(s.replace(/-/g, "+").replace(/_/g, "/") + pad, "base64"));
  },
};

export async function generateKeyPair(): Promise<KeyPair> {
  const privateKey = ed.utils.randomPrivateKey();
  const publicKey = await ed.getPublicKeyAsync(privateKey);
  return { publicKey, privateKey };
}

export function didFromPublicKey(publicKey: Uint8Array, ledger = "main"): string {
  // Multibase z prefix for base58btc would be ideal; using base64url for simplicity here
  return `did:atp:${ledger}:${b64u.encode(publicKey).slice(0, 32)}`;
}

/**
 * Create a compact JWS over the given payload using Ed25519.
 * Header: { alg: "EdDSA", typ: "JWT" }
 */
export async function signJWS(payload: object, privateKey: Uint8Array): Promise<string> {
  const header = { alg: "EdDSA", typ: "JWT" };
  const headerB64 = b64u.encode(new TextEncoder().encode(JSON.stringify(header)));
  const payloadB64 = b64u.encode(new TextEncoder().encode(JSON.stringify(payload)));
  const signingInput = `${headerB64}.${payloadB64}`;
  const sig = await ed.signAsync(new TextEncoder().encode(signingInput), privateKey);
  return `${signingInput}.${b64u.encode(sig)}`;
}

export async function verifyJWS(jws: string, publicKey: Uint8Array): Promise<boolean> {
  const [h, p, s] = jws.split(".");
  if (!h || !p || !s) return false;
  const signingInput = `${h}.${p}`;
  try {
    return await ed.verifyAsync(b64u.decode(s), new TextEncoder().encode(signingInput), publicKey);
  } catch {
    return false;
  }
}

export function hashAttestation(obj: object): string {
  // Canonical JSON (sorted keys) hashed with SHA-256
  const canonical = JSON.stringify(obj, Object.keys(obj).sort());
  return Buffer.from(sha256(new TextEncoder().encode(canonical))).toString("hex");
}

export { b64u };

File: sdks/python/atp/crypto.py
"""Cryptographic primitives — Ed25519 signing, JWS, canonical hashing.

Wire-compatible with the TypeScript SDK:
  - JWS compact serialization with header {alg: EdDSA, typ: JWT}
  - base64url encoding without padding
  - JSON canonicalization: sorted keys, no whitespace
  - SHA-256 for attestation hashing
"""
from __future__ import annotations
import base64
import hashlib
import json
from dataclasses import dataclass

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives import serialization
from cryptography.exceptions import InvalidSignature


@dataclass
class KeyPair:
    private_key: Ed25519PrivateKey
    public_key: Ed25519PublicKey

    @property
    def public_bytes(self) -> bytes:
        return self.public_key.public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

    @property
    def private_bytes(self) -> bytes:
        return self.private_key.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption(),
        )


def generate_keypair() -> KeyPair:
    priv = Ed25519PrivateKey.generate()
    return KeyPair(private_key=priv, public_key=priv.public_key())


def b64u_encode(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).decode().rstrip("=")


def b64u_decode(s: str) -> bytes:
    pad = "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s + pad)


def did_from_public_key(pubkey: bytes, ledger: str = "main") -> str:
    return f"did:atp:{ledger}:{b64u_encode(pubkey)[:32]}"


def _canonical(obj: dict) -> bytes:
    return json.dumps(obj, sort_keys=True, separators=(",", ":")).encode()


def sign_jws(payload: dict, keypair: KeyPair) -> str:
    header = {"alg": "EdDSA", "typ": "JWT"}
    header_b64 = b64u_encode(_canonical(header))
    payload_b64 = b64u_encode(_canonical(payload))
    signing_input = f"{header_b64}.{payload_b64}".encode()
    signature = keypair.private_key.sign(signing_input)
    return f"{header_b64}.{payload_b64}.{b64u_encode(signature)}"


def verify_jws(jws: str, public_key: Ed25519PublicKey) -> bool:
    parts = jws.split(".")
    if len(parts) != 3:
        return False
    signing_input = f"{parts[0]}.{parts[1]}".encode()
    try:
        public_key.verify(b64u_decode(parts[2]), signing_input)
        return True
    except InvalidSignature:
        return False
    except Exception:
        return False


def hash_attestation(obj: dict) -> str:
    return hashlib.sha256(_canonical(obj)).hexdigest()
